File: mytoolboxes/HeliCam/C4Utility/examples/Python/c4DemodSimple/c4DemodSimple.py
##
# \file        c4DemodSimple.py
# \author      Balz Marty
# \date        24 April 2025
# \copyright   Heliotis, 2025
# \version     1.0.1
#
# \brief       Simple Python example based on Harvester package for heliCam C4
import os
import numpy as np
import matplotlib.pyplot as plt
from time import sleep

from harvesters.core import Harvester
from packaging.version import parse

def selectDevice(h):
    """
    Scan for available devices and print device list
    let the user select one and open the connection to the device
    if just one device available, open it without user interaction
        param h harvester object
        return harvesters camera object
    """

    h.update()
    NDevices = len(h.device_info_list)
    print("{} device(s) detected on the Network:\n".format(NDevices))
    for i,dev in enumerate(h.device_info_list):
        print("{}) {} ({})".format(i+1,dev.id_,dev.serial_number))

    if NDevices == 1:
        selectionInt = 1
    else:
        selectionStr = input("Select a device (0=exit): ")
        selectionInt = int(selectionStr)
        if((selectionInt <= 0) or (selectionInt > NDevices)):
            exit('No device selected - exit script')

    deviceID = h.device_info_list[selectionInt-1].id_
    print('selected device:', deviceID,'\n')
    return h.create(selectionInt-1)

def cameraConfig(camera):
    """
    simple configuration example of heliCam C4 using internal reference
        param camera harvesters camera object
    """

    # Experiment Parameters


    # # Lock-In Amplification (LIA)

    # # Sensor sensitivity in %/100, 0.5 for C4-S40, 0.2 for C4-S40U, 0.05 for C4-S41U and 0.25 for C4M
    # sensitivity = 0.5
    # # Number of intergration periods
    # NPeriods = 10
    # # Background suppression on/off switch, 'AC' or 'DC'
    # coupling = 'DC'
    # # Reference frequency in Hz
    # refFrequency = 10000.0
    # # Source of reference signal, 'Internal' or 'External'
    # refSource = 'Internal'
    # # Expected frequency deviation of external reference input in %
    # expFrequencyDev = 5
    # # Number of frames to be recorded
    # NFrames = 60

    # # Illumination Driving Signal

    # # Signal generator DC offset in % of full range
    # sgnOffset = 20.0
    # # Signal generator peak-to-peak amplitude in % of full range
    # sgnAmplitude = 10.0
    # # Signal generator frequency in Hz
    # sgnFrequency = 9975.0


    # # Configuration

    # # Trigger

    # camera.remote_device.node_map.TriggerSelector.value = "RecordingStart"
    # camera.remote_device.node_map.TriggerMode.value = "Off"
    # camera.remote_device.node_map.TriggerSelector.value = "FrameStart"
    # camera.remote_device.node_map.TriggerMode.value = "On"
    # camera.remote_device.node_map.TriggerSource.value = "Software"

    # # LIA

    # camera.remote_device.node_map.DeviceOperationMode.value = "LockInCam"
    # camera.remote_device.node_map.Scan3dExtractionMethod.value = "rawIQ"

    # camera.remote_device.node_map.LockInSensitivity.value = sensitivity
    # camera.remote_device.node_map.LockInTargetTimeConstantNPeriods.value = NPeriods
    # camera.remote_device.node_map.LockInCoupling.value = coupling
    # camera.remote_device.node_map.LockInExpectedFrequencyDeviation.value = expFrequencyDev
    # camera.remote_device.node_map.LockInTargetReferenceFrequency.value = refFrequency
    # camera.remote_device.node_map.AcquisitionBurstFrameCount.value = NFrames

    # camera.remote_device.node_map.LockInReferenceSourceType.value = refSource

    # # For external reference signal only
    # camera.remote_device.node_map.LockInReferenceFrequencyScaler.value = "Off"
    # camera.remote_device.node_map.LockInReferenceSourceSignal.value = "FI2"

    # # Illumination

    # camera.remote_device.node_map.SignalGeneratorOffset.value = sgnOffset
    # camera.remote_device.node_map.SignalGeneratorAmplitude.value = sgnAmplitude
    # camera.remote_device.node_map.LightControllerSelector.value = "LightController0"
    # camera.remote_device.node_map.SignalGeneratorModulationMode.value = "On"
    # camera.remote_device.node_map.SignalGeneratorFrequency.value = sgnFrequency
    # camera.remote_device.node_map.LightControllerSource.value = "SignalGenerator"

# Lock-In Amplification (LIA)

    # Sensor sensitivity in %/100, 0.5 for C4-S40, 0.2 for C4-S40U, 0.05 for C4-S41U and 0.25 for C4M
    sensitivity = 1
    # Number of intergration periods
    NPeriods = 20
    # Background suppression on/off switch, 'AC' or 'DC'
    coupling = 'DC'
    # Reference frequency in Hz
    refFrequency = 5000.0
    # Source of reference signal, 'Internal' or 'External'
    refSource = 'External'
    # Expected frequency deviation of external reference input in %
    expFrequencyDev = 1
    # Number of frames to be recorded
    NFrames = 4

    # Illumination Driving Signal

    # Signal generator DC offset in % of full range
    sgnOffset = 20.0
    # Signal generator peak-to-peak amplitude in % of full range
    sgnAmplitude = 10.0
    # Signal generator frequency in Hz
    sgnFrequency = 9975.0


    # Configuration

    # Trigger

    camera.remote_device.node_map.TriggerSelector.value = "RecordingStart"
    camera.remote_device.node_map.TriggerMode.value = "Off"
    camera.remote_device.node_map.TriggerSelector.value = "FrameStart"
    camera.remote_device.node_map.TriggerMode.value = "On"
    camera.remote_device.node_map.TriggerSource.value = "FI2"

    # LIA

    camera.remote_device.node_map.DeviceOperationMode.value = "LockInCam"
    camera.remote_device.node_map.Scan3dExtractionMethod.value = "rawIQ"

    camera.remote_device.node_map.LockInSensitivity.value = sensitivity
    camera.remote_device.node_map.LockInTargetTimeConstantNPeriods.value = NPeriods
    camera.remote_device.node_map.LockInCoupling.value = coupling
    camera.remote_device.node_map.LockInExpectedFrequencyDeviation.value = expFrequencyDev
    camera.remote_device.node_map.LockInReferenceFrequencyScaler.value = 'DivideBy4'
    camera.remote_device.node_map.LockInTargetReferenceFrequency.value = refFrequency
    camera.remote_device.node_map.AcquisitionBurstFrameCount.value = NFrames

    camera.remote_device.node_map.LockInReferenceSourceType.value = refSource

    # For external reference signal only
    camera.remote_device.node_map.LockInReferenceFrequencyScaler.value = "On"
    camera.remote_device.node_map.LockInReferenceSourceSignal.value = "FI2"

    # Illumination

    camera.remote_device.node_map.SignalGeneratorOffset.value = sgnOffset
    camera.remote_device.node_map.SignalGeneratorAmplitude.value = sgnAmplitude
    camera.remote_device.node_map.LightControllerSelector.value = "LightController0"
    camera.remote_device.node_map.SignalGeneratorModulationMode.value = "On"
    camera.remote_device.node_map.SignalGeneratorFrequency.value = sgnFrequency
    camera.remote_device.node_map.LightControllerSource.value = "SignalGenerator"

def getOutputShape(camera):
    """
    Get shape of measurement's in-phase/quadrature lock-in output.
        param h harvester object
        return tuple (NChannels,Nframes,height,width)
    """
    NFrames = camera.remote_device.node_map.AcquisitionBurstFrameCount.value
    height = camera.remote_device.node_map.Height.value
    width = camera.remote_device.node_map.Width.value

    return 2,NFrames,height,width

def acquire(camera,timeout=30):
    """
    Initiate a measurement and retrieve lock-in data.
        param h harvester object
        param t float
        return numpy array [IRaw,QRaw]
    """

    outputShape = getOutputShape(camera)

    camera.start()
    camera.remote_device.node_map.TriggerSelector.value = 'FrameStart'
    camera.remote_device.node_map.TriggerSoftware.execute()

    with camera.fetch(timeout=timeout) as buffer:

        data = np.array([img.data%2**15//4 for img in buffer.payload.components]).reshape(outputShape)

    camera.stop()

    return data

def plotData(I, Q):
    """
    Plot measurement data
    """

    _,height,width=I.shape
    
    coords = [[110, 120],[451, 454]]
    IPlot = np.array([I[:,coord[1],coord[0]] for coord in coords])
    QPlot = np.array([Q[:,coord[1],coord[0]] for coord in coords])
        
    rms = np.sqrt(np.mean(I**2 + Q**2, axis=0))

    # styling for the I and Q signals
    style = {"markevery": 1, "mec": "1.0", "ls": ":", "marker": "o"}
    fig = plt.figure(figsize=(12, 9))
    gs = fig.add_gridspec(2, 2)

    ax = fig.add_subplot(gs[0, 0])
    ax.plot(IPlot[0, :], "C1", **style)
    ax.plot(QPlot[0, :], "C2", **style)
    ax.set_title(f"Raw I/Q Signal at (x0={coords[0][0]}, y0={coords[0][1]})")
    ax.set_xlabel("Frame Nr.")
    ax.set_ylabel("Intensity")
    ax.legend(["I", "Q"], loc=4)

    ax = fig.add_subplot(gs[1, 0])
    ax.plot(IPlot[1, :], "C3", **style)
    ax.plot(QPlot[1, :], "C4", **style)

    ax.set_title(f"Raw I/Q Signal at (x1={coords[1][0]}, y1={coords[1][1]})")
    ax.set_xlabel("Frame Nr.")
    ax.set_ylabel("Intensity")
    ax.legend(["I", "Q"], loc=4)

    ax = fig.add_subplot(gs[:, 1])
    im = ax.pcolormesh(rms , cmap="inferno")#,vmin=0, vmax=500
    fig.colorbar(im, ax=ax)
    ax.hlines(coords[0][1],
              xmin=0,
              xmax=width,
              colors="k",
              ls="-.",
              linewidth=0.75)
    ax.vlines(coords[0][0],
              ymin=0,
              ymax=height,
              colors="k",
              ls="-.",
              linewidth=0.75,
              label="_nolegend_")
    ax.hlines(coords[1][1],
              xmin=0,
              xmax=width,
              colors="royalblue",
              ls="-.",
              linewidth=0.75)
    ax.vlines(coords[1][0],
              ymin=0,
              ymax=height,
              colors="royalblue",
              ls="-.",
              linewidth=0.75,
              label="_nolegend_")
    marker = {"marker": "x", "markeredgewidth": 1, "markersize": 8}
    ax.plot(*coords[0], "k", **marker)
    ax.plot(*coords[1], "royalblue", **marker)
    ax.set_title(f"Root Mean Square Lock-In Amplitude")
    ax.legend(["(x0, y0)", "(x1, y1)"], loc=4)
    ax.set_xlabel("X")
    ax.set_ylabel("Y")

    # set the spacing between subplots
    fig.tight_layout(pad=2)
    plt.subplots_adjust(left=None, bottom=None, right=None, top=0.85, wspace=None, hspace=None)
    plt.suptitle("Lock-In Amplification with Sinusoid Input")
    plt.show()

if __name__ == '__main__':

    print('********************************************************')
    print('* Configure and acquire data in "Raw I/Q Mode"')
    print('********************************************************')

    h = Harvester()
    # read heliotis *.CTI path from environment variable
    ctiFile = os.environ['DIAPHUS_GENTL64_FILE']
    h.add_file(ctiFile)

    # create camera object for interaction
    camera = selectDevice(h)

    # configure camera
    cameraConfig(camera)
    print(f"LockInActualReferenceFrequency: {camera.remote_device.node_map.LockInActualReferenceFrequency.value}Hz")
    sleep(2.5) # make sure illumination output is settled

    # acquire data
    rawI,rawQ = acquire(camera)

    # Discard initial frames lacking information present in firmware versions <= 1.9.2
    NFramesDiscard = 0 if parse(camera.remote_device.node_map.DeviceFirmwareVersion.value) > parse('1.9.2') else 2

    I = rawI[NFramesDiscard:].astype(float)
    Q = rawQ[NFramesDiscard:].astype(float)

    # (Optionally) remove the offset (simple time average per pixel)
    removeOffset = True

    if removeOffset:

        I -= np.mean(I, axis=0)
        Q -= np.mean(Q, axis=0)


    # plot measurement data
    plotData(I,Q)

    # free resources
    camera.stop()
    camera.destroy()

File: mytoolboxes/HeliCam/c4DemodSimple.m
clc; clear all; close all;

if ~NET.isNETSupported
    disp('Supported .NET Framework not found')
    return
end

NET.addAssembly('C4HdlCLR');
import C4HdlCLR.*

disp('*************************************************')
disp('* Matlab example c4DemodSimple based on C4HdlCLR *')
disp('*************************************************')

% establish camera connection
c4sys = heliotis.C4HandlerCLR();
c4sys.reset();
ifNo = selectInterface(c4sys);
c4if = c4sys.openInterface(ifNo);
devNo = selectDevice(c4if);
c4dev = c4if.openDevice(devNo);

% configure camera and wait for LED output
fprintf('initializing camera...\n')
initializeDevice(c4dev)
pause(3) %in seconds
fprintf('... done')

% enter camera acquisition mode
c4dev.startAcquisition(4)

% launch frame start trigger
c4dev.executeCommand("TriggerSoftware");

% load measurement data from buffer
c4buf = c4dev.getBuffer(10000);

% retrieve rawI and rawQ component
NFrames = c4buf.readInteger("ChunkPartCount")/2;
frameDimension = c4buf.getPartDimension(1);
width = frameDimension(1);
height = frameDimension(2);

rawI = zeros(NFrames,height,width);
rawQ = zeros(NFrames,height,width);

for i=0:NFrames-1

    rawI(i+1,:,:) = transpose(reshape(uint16(c4buf.getDataPartUint16(i)),width,height));
    rawQ(i+1,:,:) = transpose(reshape(uint16(c4buf.getDataPartUint16(i+NFrames)),width,height));

end

rawI = (mod(rawI, 2^15)) / 2^2;
rawQ = (mod(rawQ, 2^15)) / 2^2;

% Discard initial frames lacking information
NFramesDiscard = 3;

I = rawI(NFramesDiscard:NFrames,:,:);
Q = rawQ(NFramesDiscard:NFrames,:,:);

% (Optionally) remove the offset (simple time average per pixel)
removeOffset = true;

if removeOffset

    I = I - mean(I, 1);
    Q = Q - mean(Q, 1);

end

% plot measurement data
plotData(I,Q)

% release buffer
c4buf.release();

% stop acquisition
c4dev.stopAcquisition();

% release resources
c4dev.release()
c4if.release()

function ifNo = selectInterface(c4sys)
    % interface selection with user interaction
    ifNo = -1;
    disp('Interfaces: ')
    nofIf = c4sys.updateInterfaceList();
    if (nofIf == 0)
        disp('No interface detected!');
        return
    end
    for i = 0:(nofIf-1)
        curIfName = c4sys.getInterfaceName(i);
        disp(num2str(i + 1) + ": " + string(curIfName));
    end
    val = input('select an interface (exit with any other value): ');
    if ((val <= 0) || (val > nofIf))
        disp('Interface selection is out of range!');
        return
    end
    ifNo = (val - 1);
end

function devNo = selectDevice(c4if)
    % device selection with user interaction 
    devNo = -1;
    disp('Device: ')
    nofDev = c4if.updateDeviceList();
    if (nofDev == 0)
        disp('No device detected!');
        return
    end
    for i = 0:(nofDev-1)
        curDevName = c4if.getDeviceName(i);
        disp(num2str(i + 1) + ": " + string(curDevName));
    end
    val = input('select a device (exit with any other value): ');
    if ((val <= 0) || (val > nofDev))
        disp('Device selection is out of range!');
        return
    end
    devNo = (val - 1);
end

function initializeDevice(c4dev)
    % simple configuration example of heliCam C8 using internal reference
    % signal
    
    % Experiment Parameters


    % Lock-In Amplification (LIA)

    % Sensor sensitivity in %/100, 0.5 for C4-S40, 0.2 for C4-S40U, 0.05 for C4-S41U and 0.25 for C4M
    sensitivity = 0.5;
    % Number of intergration periods
    NPeriods = 10;
    % Background suppression on/off switch, 'AC' or 'DC'
    coupling = 'DC';
    % Reference frequency in Hz
    refFrequency = 10000.0;
    % Source of reference signal, 'Internal' or 'External'
    refSource = 'Internal';
    % Expected frequency deviation of external reference input in %
    expFrequencyDev = 5;
    % Number of frames to be recorded
    NFrames = 60;

    % Illumination Driving Signal

    % Signal generator DC offset in % of full range
    sgnOffset = 20.0;
    % Signal generator peak-to-peak amplitude in % of full range
    sgnAmplitude = 10.0;
    % Signal generator frequency in Hz
    sgnFrequency = 9975.0;


    % Configuration

    % Trigger

    c4dev.writeString("TriggerSelector", "RecordingStart");
    c4dev.writeString("TriggerMode", "Off");
    c4dev.writeString("TriggerSelector", "FrameStart");
    c4dev.writeString("TriggerMode", "On");
    c4dev.writeString("TriggerSource", "Software");

    % LIA

    c4dev.writeString("DeviceOperationMode", "LockInCam");
    c4dev.writeString("Scan3dExtractionMethod", "rawIQ");

    c4dev.writeFloat("LockInSensitivity", sensitivity);
    c4dev.writeInteger("LockInTargetTimeConstantNPeriods", NPeriods);
    c4dev.writeString("LockInCoupling", coupling);
    c4dev.writeInteger("LockInExpectedFrequencyDeviation", expFrequencyDev);
    c4dev.writeFloat("LockInTargetReferenceFrequency", refFrequency);
    c4dev.writeInteger("AcquisitionBurstFrameCount", NFrames);

    c4dev.writeString("LockInReferenceSourceType", refSource);

    % For external reference signal only
    c4dev.writeString("LockInReferenceFrequencyScaler", "Off");
    c4dev.writeString("LockInReferenceSourceSignal", "FI2");

    % Illumination

    c4dev.writeFloat("SignalGeneratorOffset", sgnOffset);
    c4dev.writeFloat("SignalGeneratorAmplitude", sgnAmplitude);
    c4dev.writeString("LightControllerSelector", "LightController0");
    c4dev.writeString("SignalGeneratorModulationMode", "On");
    c4dev.writeFloat("SignalGeneratorFrequency", sgnFrequency);
    c4dev.writeString("LightControllerSource", "SignalGenerator");


    % Read feature values from camera

    ActualRefFrequency = c4dev.readFloat("LockInActualReferenceFrequency");

    fprintf('Actual Reference Frequency :  %0.2f Hz\n',ActualRefFrequency)

end

function plotData(I,Q)
    % plot the measured data


    % plot time series of selected pixels

    xs = [100, 454];
    ys = [100, 412];
    cs = ['r','b'];
    
    NFrames = size(I,1);
    
    for i = 1:length(xs)
    
        subplot(2,2,i);
        plot(1:NFrames,I(:,xs(i),ys(i)),Color=cs(i),LineStyle="-")
        hold on
        plot(1:NFrames,Q(:,xs(i),ys(i)),Color=cs(i),LineStyle=":")
        hold on
        
        legend('I','Q')
        xlabel('frame number')
        ylabel('I/Q')
        title(['point ' num2str(i) ' (x=' num2str(xs(i)) ',y=' num2str(ys(i)) ')'])
        
    
    end
    

    % plot root mean square (rms) amplitude across field of view

    %rms = squeeze(sqrt(mean(I.^2 + Q.^2)));
    rms = squeeze(mean(I));

    subplot(2,2,[3,4])
    
    imagesc(rms)
    colorbar
    hold on
    
    for i = 1:length(xs)
    
        plot(xs(i),ys(i),Color=cs(i),Marker='+',MarkerSize=15,LineWidth=2);
        hold on
    
    end
    
    xlabel('x in pxl')
    ylabel('y in pxl')
    title('rms amplitude')
    
    hold off

end